softmax: Triton kernel for BSR inputs

nikitaved opened the pull request:
Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101978
* #103548
* #103544
* #103750
* #103835
* __->__ #102095



cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer

nikitaved committed:
softmax: Triton kernel for BSR inputs

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"




[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -3353,6 +3353,45 @@ def setUp(self):
+            input_tiled.copy_(dense)
+
+            res_tri = bsr_softmax(bsr)
+            res_dense = input.softmax(-1)
```
I'd expect the reference to be https://pytorch.org/docs/master/generated/torch.sparse.softmax.html#torch.sparse.softmax since we're really following masked semantics here if this softmax is to be used in the context of SDPA.

Is that correct?

nikitaved replied:
`torch.sparse` is limited to non-batched COO only. Besides, the dense reference is masked to produce zeroes after softmax.

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on torch/sparse/_triton_ops.py:
```diff
@@ -703,6 +703,102 @@ def bsr_dense_mm(
+            max_row_nnz = triton.next_power_of_2(max_row_nnz)
+
+        crow_indices = input.crow_indices().unsqueeze(0).flatten(0, -2)
+        values = input.values().transpose(-3, -2).contiguous().unsqueeze(0).flatten(0, -4).reshape(-1, row_block, nnz * col_block)
```
TODO: add a guard against accidental in-place storage modification.

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -3353,6 +3353,53 @@ def setUp(self):
+        for row_block, col_block, b, m, n in itertools.product(block_size, block_size, batches, size, size):
+            input = tensor(b + (m, n))
+            input.diagonal(dim1=-2, dim2=-1).fill_(m * n)
+            input = self._to_block_triangular_inplace(input, row_block, col_block)
```
Why specifically block triangular?

nikitaved replied:
I think it is a very nice deterministic test case will all rows having distinct nnz.

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on torch/sparse/_triton_ops.py:
```diff
@@ -703,6 +703,102 @@ def bsr_dense_mm(
+        if row_nnz == 0:
+            return
+
+        row_arange = tl.arange(0, MAX_ROW_NNZ)
```
Only a single row is loaded, not a block row, so the kernel becomes almost identical to the softmax example in the Triton tutorial.

nikitaved replied:
All right, so there is a cap in this kernel:
```python
In [1]: import torch                                                                                                                                                                                                                          
                                                                                                                                                                                                                                              
In [2]: x = torch.rand(1, 150000, device='cuda', dtype=torch.half)                                                                                                                                                                            
                                                                                                                       
In [3]: from torch.sparse._triton_ops import bsr_softmax                                                                                                                                                                                      
                                                           
In [4]: res1 = bsr_softmax(x.to_sparse_bsr(x.shape))                                                                   
<ipython-input-4-33c261e075e6>:1: UserWarning: Sparse BSR tensor support is in beta state. If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues. (Triggered 
internally at /home/nik/git/Quansight/pytorch/aten/src/ATen/SparseCsrTensorImpl.cpp:53.)                                                                                                                                                      
  res1 = bsr_softmax(x.to_sparse_bsr(x.shape)) 
...
CompilationError: at 30:30:        + crow_indices_stride * row_block_pid
    )
    nnz_offset = tl.load(crow_indices_offset_ptr)
    nnz_offset_next = tl.load(crow_indices_offset_ptr + crow_indices_stride)

    # Compute nnz for the row with number row_block_pid.
    # If it is zero, skip the row.
    row_nnz = nnz_offset_next - nnz_offset
    if row_nnz == 0:
        return

    row_arange = tl.arange(0, MAX_ROW_NNZ)
                              ^
ValueError('numel (262144) exceeds triton maximum tensor numel (131072)')

```

cpuhrsch replied:
@nikitaved - If it's a straightforward modification, I think it's worth it. Otherwise we can do it as a follow up.

Overall it's unlikely we'll see inner dimensions of that size in the context of SDPA.

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "softmax: Triton kernel for BSR inputs"

Implements `softmax` Triton kernel for BSR inputs. So far, only over `dim=-1`.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge
